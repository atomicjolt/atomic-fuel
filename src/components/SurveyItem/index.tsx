import React from 'react';
import { Link } from 'react-router-dom';
import i18n from 'i18next';
import Tippy from '@tippyjs/react';
import { withSettings } from 'atomic-fuel';

import QualtricsLink from '../common/qualtrinks_link';
import Menu from '../Menu';
import Button, { ButtonType } from '../Button';
import Publish from '../Publish';
import { SurveyAssignment, Settings } from '../../types';

import './styles.scss';

interface Props {
  readonly surveyAssignment: SurveyAssignment,
  readonly onDelete: (survey: SurveyAssignment) => void,
  readonly settings: Settings,
}

function SurveyItem({ surveyAssignment, onDelete, settings }: Props) {
  const createdAt = new Date(surveyAssignment.createdAt).toLocaleDateString();
  const dueDate = surveyAssignment.dueAt
    ? new Date(surveyAssignment.dueAt).toLocaleDateString()
    : '-';
  const { canvas_url: canvasUrl, canvas_course_id: canvasCourseId } = settings;
  const assignmentUrl = `${canvasUrl}/courses/${canvasCourseId}/assignments/${surveyAssignment.lmsAssignmentId}`;

  return (
    <tr className="assignment-item">
      <td className="assignment-item__name">
        <a href={assignmentUrl} target="_top">
          {surveyAssignment.name}
        </a>
      </td>
      <td className="assignment-item__due">{dueDate}</td>
      <td className="assignment-item__date">{createdAt}</td>
      <td className="assignment-item__completed">{surveyAssignment.completed}</td>
      <td className="assignment-item__actions">
        <QualtricsLink
          to="results"
          id={surveyAssignment.survey.qSurveyId}
          canAccess={surveyAssignment.survey.canAccess}
        >
          Results
        </QualtricsLink>
        { surveyAssignment.survey.owns ?
          <Publish
            published={surveyAssignment.published}
            surveyId={surveyAssignment.id}
          /> :
          <Tippy
            animation="shift-away-subtle"
            content="Cannot Publish a Shared Survey."
            duration={0}
            placement="bottom"
          >
            <i className="material-icons-outlined locked">lock</i>
          </Tippy>}
        <Menu renderChildren={(
          activeClass,
          _,
          menuOpen,
          menuRef,
          onKeyPress,
          toggleMenu,
        ) => {
          const ariaOptions = {
            'aria-label': i18n.t('assignment options'),
            'aria-haspopup': 'menu',
            'aria-expanded': menuOpen ? 'true' : 'false',
          };

          return (
            <div ref={menuRef} className={`aj-menu-container ${activeClass}`}>
              <Tippy
                animation="shift-away-subtle"
                content="Survey Options"
                duration={0}
                placement="bottom"
              >
                <Button
                  ariaOptions={ariaOptions}
                  color="gray"
                  buttonType={ButtonType.icon}
                  onClick={toggleMenu}
                >
                  <i className="material-icons-outlined" arria-hidden="true">more_vert</i>
                </Button>
              </Tippy>
              <ul
                className="aj-menu"
                role="menu"
                onClick={(e) => e.stopPropagation()}
                onKeyPress={onKeyPress}
              >
                <li>
                  <QualtricsLink
                    to="edit"
                    id={surveyAssignment.survey.qSurveyId}
                    canAccess={surveyAssignment.survey.canAccess}
                  >
                    <i className="material-icons-outlined">mode_edit</i>
                    Edit Content
                  </QualtricsLink>
                </li>
                <li>
                  <Link to={`/${surveyAssignment.id}/settings`}>
                    <i className="material-icons-outlined">settings</i>
                    Settings
                  </Link>
                </li>
                <li>
                  <button onClick={() => onDelete(surveyAssignment)}>
                    <i className="material-icons-outlined">delete</i>
                    Delete Survey
                  </button>
                </li>
              </ul>
            </div>
          );
        }}
        />
      </td>
    </tr>
  );
}

export default withSettings(SurveyItem);
