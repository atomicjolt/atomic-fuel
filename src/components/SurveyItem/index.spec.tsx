import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import SurveyItem from '.';


describe('SurveyItem component', () => {
  defaultSettings({
    canvas_url: 'url',
    canvas_course_id: 1234,
  });

  it('Should match default snapshot', () => {
    const { asFragment } = render(
      <MockedProvider>
        <SurveyItem
          surveyAssignment={{
            id: '11',
            anonymousGrading: false,
            lmsAssignmentId: '1234',
            name: 'name',
            createdAt: '2020/12/12',
            dueAt: '2020/12/12',
            published: false,
            scoreMaximum: 10,
            completed: 15,
            responses: [],
            survey: {
              id: 'string',
              qSurveyId: 'id',
              createdAt: '2020/12/12',
              published: false,
              canAccess: true,
              owns: true,
            },
          }}
          onDelete={() => {}}
        />
      </MockedProvider>,
      { wrapper: MemoryRouter },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
