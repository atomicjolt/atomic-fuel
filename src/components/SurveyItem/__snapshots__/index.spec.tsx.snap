// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyItem component Should match default snapshot 1`] = `
<DocumentFragment>
  <a
    href="undefined/courses/1234/assignments/1234"
    target="_top"
  >
    name
  </a>
  12/12/202012/12/202015
  <a
    class="qualtrics-link qualtrics-link__results"
    href="https://ausurveys.co1.qualtrics.com/results/?surveyId=id"
    rel="noopener noreferrer"
    target="_blank"
  >
    Results
  </a>
  <button
    class="aj-btn aj-btn--gray publish-btn unpublished"
    type="button"
  >
    <i
      class="material-icons"
    >
      do_not_disturb
    </i>
  </button>
  <div
    class="aj-menu-container "
  >
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      class="aj-btn aj-btn--gray aj-btn--icon"
      type="button"
    >
      <i
        arria-hidden="true"
        class="material-icons-outlined"
      >
        more_vert
      </i>
    </button>
    <ul
      class="aj-menu"
      role="menu"
    >
      <li>
        <a
          class="qualtrics-link qualtrics-link__edit"
          href="https://ausurveys.co1.qualtrics.com/surveys/id/edit"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="material-icons-outlined"
          >
            mode_edit
          </i>
          Edit Content
        </a>
      </li>
      <li>
        <a
          href="/11/settings"
        >
          <i
            class="material-icons-outlined"
          >
            settings
          </i>
          Settings
        </a>
      </li>
      <li>
        <button>
          <i
            class="material-icons-outlined"
          >
            delete
          </i>
          Delete Survey
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
